autolab/Autolab
# Watchlist entries

## Description
This PR uses the endpoint integrations from https://github.com/autolab/Autolab/pull/1262 to complete the watchlist entries front-end for the new metrics feature. It combines instances for each student into a single watchlist row (i.e. each tab has at most one row for each student). The code for generating the watchlist is moved into a new watchlist.js file.
<img width="1184" alt="Screen Shot 2020-12-11 at 10 53 48 PM" src="https://user-images.githubusercontent.com/13857201/101973688-da4f4d80-3c07-11eb-995b-08d74546c8b5.png">
<img width="1118" alt="Screen Shot 2020-12-11 at 10 53 30 PM" src="https://user-images.githubusercontent.com/13857201/101973691-dde2d480-3c07-11eb-8ecc-17a002b465d8.png">
<img width="979" alt="Screen Shot 2020-12-12 at 1 19 08 AM" src="https://user-images.githubusercontent.com/13857201/101976934-0a065180-3c18-11eb-9de0-69a9a4dee5d3.png">

## Motivation and Context
Completes the watchlist front-end for the new metrics feature (except for the refresh button), with API endpoints integrated.



## How Has This Been Tested?

This is one particular flow that I tried:

1. Set up risk metrics - I set metrics that would lead to a large number of watchlist instances (1 grace day, 1% grade drop for 2 consecutive assignments, 1 no submission, 1 submitted assignment below 100%
2. Go to the watchlist tab -> make sure students are appearing in the `new` tab
3. Click on any student's outgoing gradebook link -> make sure that it links to the correct student's grades
4. Hover over the condition tags -> make sure the correct information about assignments appears
5. Click the contact button for any student -> make sure that it links to a mail client with the correct student email filled in (the browser's default email client has to be set for the mailto to work - if you're using Chrome, you can do that [here](chrome://settings/handlers))
6. Go back to the watchlist page -> make sure the contacted student is no longer in the `new` tab and has moved to `contacted`
7. In the `new` tab, click the resolve button for any student -> make sure the resolved student is no longer in the `new` tab and has moved to `resolved`
8. Check the boxes for a few students and click the top-level contact button -> check that the correct emails are listed in the mailto and that the students are moved into the `contacted` tab
9. Check the boxes for a few students and click the top-level resolve button -> check that the students are moved into the `resolved` tab
10. Resolve a student in the `contacted` tab -> make sure the student moves to the `resolved` tab
11. In the `new` tab, click the top-level select-all checkbox and click contact -> check that the correct emails are listed in the mailto and that all students are moved into the `contacted` tab (there should be no students in `new` now)
12. In the `contacted` tab, click the top-level select-all checkbox and click resolve -> check that all students are moved into the `resolved` tab (there should be no students in `contacted` now)
13. Return to the main course page -> the metrics button should no longer have a notification badge
14. Change a risk metric and return to the watchlist -> there should now be entries in the `archived` tab, and students in other tabs who also appear in `archived` should have an exclamation icon in the row
15. Additionally, check that the instructor cannot attempt to contact students from any tab besides the `new` tab and cannot attempt to resolve students from any tab besides the `new` and `contacted` tabs.





## Types of changes


## Checklist:


## Other issues / help required
Since the watchlist combines instances for a particular student into a single row, the metrics notification badge - which currently counts the number of instances - can be misleading. We may want to change the endpoint for getting the number of  new instances to instead get the number of students with new instances.
Additionally, the refresh button still needs to update the "last updated" date and use throttle/debounce to prevent too many refreshes. 
I'd also appreciate any feedback on the appearance, especially the gradebook link icon and the archived student icon, as these were not specified in the original design. The gradebook link icon was included so that the student name could be a label on the checkbox (allowing the user to click on the name in order to check the box).




If unsure, feel free to submit first and we'll help you along.
